include/exadg/functions_and_boundary_conditions/interface_coupling.h

```diff
@@ -23,433 +23,53 @@
-
-  // sender side
-  std::map<unsigned int, std::vector<Point<spacedim>>> relevant_remote_points_per_process;
-  std::map<unsigned int, std::vector<unsigned int>>    relevant_remote_points_count_per_process;
-};
-
-
 template<int dim, int n_components, typename Number>
 class InterfaceCoupling
 {
 private:
   static unsigned int const rank =
     (n_components == 1) ? 0 : ((n_components == dim) ? 1 : numbers::invalid_unsigned_int);
 
-  typedef InterfaceCoupling<dim, n_components, Number> This;
-
-  typedef LinearAlgebra::distributed::Vector<Number> VectorType;
-  typedef LinearAlgebra::distributed::Vector<double> VectorTypeDouble;
-  typedef FaceIntegrator<dim, n_components, Number>  Integrator;
-  typedef std::pair<unsigned int, unsigned int>      Range;
-
-  typedef std::map<types::boundary_id, std::shared_ptr<FunctionCached<rank, dim, double>>>
-    MapBoundaryCondition;
+  using VectorType = LinearAlgebra::distributed::Vector<Number>;
+  using Integrator = FaceIntegrator<dim, n_components, Number>;
 
-  typedef unsigned int quad_index;
-  typedef unsigned int mpi_rank;
+  using MapBoundaryCondition =
+    std::map<types::boundary_id, std::shared_ptr<FunctionCached<rank, dim, double>>>;
 
-  typedef std::tuple<unsigned int /*face*/, unsigned int /*q*/, unsigned int /*v*/> Id;
+  using quad_index = unsigned int;
 
-  typedef std::map<Id, types::global_dof_index> MapIndex;
+  using Id = std::tuple<unsigned int /*face*/, unsigned int /*q*/, unsigned int /*v*/>;
 
-  typedef std::pair<std::vector<types::global_dof_index>, std::vector<double>> Cache;
+  using MapIndex = std::map<Id, types::global_dof_index>;
 
-  typedef std::vector<Point<dim>>                             ArrayQuadraturePoints;
-  typedef std::vector<std::vector<Cache>>                     ArrayVectorCache;
-  typedef std::vector<std::vector<Tensor<rank, dim, double>>> ArrayVectorTensor;
+  using ArrayQuadraturePoints = std::vector<Point<dim>>;
+  using ArrayVectorTensor     = std::vector<std::vector<Tensor<rank, dim, double>>>;
 
 public:
-  InterfaceCoupling() : dof_index_dst(0), dof_handler_src(nullptr), mapping_src(nullptr)
+  InterfaceCoupling() : dof_handler_src(nullptr)
   {
   }
 
   void
-  setup(std::shared_ptr<MatrixFree<dim, Number>>  matrix_free_dst_in,
-        unsigned int const                        dof_index_dst_in,
+  setup(std::shared_ptr<MatrixFree<dim, Number>>  matrix_free_dst,
+        unsigned int const                        dof_index_dst,
         std::vector<quad_index> const &           quad_indices_dst_in,
         MapBoundaryCondition const &              map_bc_in,
-        std::shared_ptr<Triangulation<dim>> const triangulation_src_in,
+        std::shared_ptr<Triangulation<dim>> const triangulation_src,
         DoFHandler<dim> const &                   dof_handler_src_in,
-        Mapping<dim> const &                      mapping_src_in,
+        Mapping<dim> const &                      mapping_src,
```

Comment:
I think we should use consistent names here in terms of the suffix `_in` or not.

@nfehn a question regarding the argument `triangulation_src`: This seems redudant as we also pass in `dof_handler_src_in`, which, according to the name and its use in FSI `driver.cpp`, must necessarily point to the same `triangulation` as the one that gets passed in. So I would opt for not repeating the same information through different channels and query it via `dof_handler_src_in.get_triangulation()` (not necessarily in this PR).

Reply:
> I think we should use consistent names here in terms of the suffix _in or not.

I just renamed those variables that are not stored in the class in the style of `variable = variable_in`.

Reply:
> a question regarding the argument triangulation_src

I have removed it.

Reply:
Here, I would tend to agree with @kronbichler to consistently write `_in` (then also for the last argument `tolerance`).

I think technically you are correct that triangulation is not needed as argument and can be queried via the dof_handler. I think I used this as an argument because the data structures are also used in the sense of mesh = {tria, mapping}, and not only dof_handler = {tria, finite_element}. So the code is somewhat clearer with separate arguments in my opinion. Overall, I am not fully convinced regarding dealii's design choice to query the triangulation via the dof_handler, but this would have to be discussed elsewhere (where my opinion will be of no relevance ... In general, I am of the opinion that we do not have to repeat design choices from dealii just because dealii decided to do it like this.). It appears somewhat arbitrary to me that dealii queries tria via dof_handler, but always uses mapping as a separate parameter. As we currently do not strive for making dealii "replacable" within ExaDG, I can live with removing this parameter.

Reply:
The main problem I see is that the mesh underlying the `DoFHandler` can be incompatible with the triangulation explicitly handed in, causing the risk for errors in case one accidentally hands in something different. We could of course check that again, but at what cost? I agree that choices do not have to be repeated, but here it would merely make our lives more complicated due to repeating code that needs to be maintained.

The reason why `DoFHandler` keeps a pointer to the triangulation is partly historical as deal.II evolved ~20 years ago. The more important reason is that `DoFHandler` in itself would merely be a storage location for numbers of the unknowns on cells, but those numbers make no sense without the cells (or faces/lines/vertices) provided by the triangulation, so `get_triangulation` is a response to the underlying grid to interpret numbers stored in `DoFHandler`. The same could be said about a `Mesh` or its fundamental ingredient `Mapping`, but here the case is not as straight-forward because
- Non-bilinear geometries were not a central design goal 20 years ago and came around later (even though before I started using deal.II), so it came by as an "addition" to existing data structures.
- The basic mappings do not *store* anything related to the mesh, but rather compute the necessary parts on demand, thus taking a different role than a `DoFHandler`, as they can be fed by an individual `cell` and do not need a `Triangulation` to be useful (the correct term would be a "mesh").
- There is sometimes a desire to create "different" mappings as one performs different steps in a simulation (e.g. two different polynomial degrees of the geometry to adhere to different accuracies of quadrature formulas), and sometimes one might be tempted to even use the same `DoFHandler` with those different mappings (albeit one is likely in the territory of variational crimes).

Either way, I see no *clean* solution that meets the typical requirements and would provide a clear advantage over the pair `Mapping, DoFHandler` (the latter implicitly containing `Triangulation`). The combination is error prone mostly because deal.II also allows users not to provide the mapping argument in several contexts, thus one can miss out these parts. Always using three arguments `Triangulation, Mapping, DoFNumbers` (I intentionally use a different name) or `Mesh, DoFHandler` (redundant triangulation; can be kept in sync with some signal mechanisms) or a super data structure `MeshDoFHandler` (consisting of `Triangulation, Mapping, DoFHandler`; duplicating some of the content, especially in more complicated problems) all have their downsides, too.

Reply:
I agree that we made the most reasonable decision here. But don't you think DoFHandler is also kind of a super data structure, given that DoFHandler itself provides cell iterator functions for example (as another example, we already discussed the MG functionality as part of DoFHandler elsewhere)? Where is "single-responsibility" here? I wonder whether one could have been more strict about what is functionality belonging to triangulation and/or to "dof numbers". To provide just one more example, consider the public function `DoFHandler::get_communicator()`, where the documentation says "Return MPI communicator used by the underlying triangulation. ". In my opinion this is simply inconsequent, independently of the design choices made some time ago and discussed above. To keep interfaces lean, one should enforce `dof_handler.get_triangulation().get_communicator()` instead of such short cuts that blur "single responsibility".

> numbers make no sense without the cells (or faces/lines/vertices) provided by the triangulation, so get_triangulation is a response to the underlying grid to interpret numbers stored in DoFHandler. 

This sounds like a solution that you are describing here, i.e. `triangulation` has to be passed to public functions of `DoFHandler` where information on cells/faces/etc is needed, but no pointer to the triangulation is stored, at least no public getter `get_triangulation()`. In this sense, the dof-handler would be some kind of helper/utility functionality (and if data needs to be stored, it has to be enforced in debug mode that one is using the correct data structures).